### ISSUE TITLE
API: prettier.format reloads plugins on every call

### ISSUE THREAD

**Issue by velut** [prettier/prettier#7229]
**Description**

Calling the `prettier.format` Node.js API causes plugins to be reloaded every time, even if the formatting options remain the same.
This appears to slow down formatting, since plugin directories are (seemingly) accessed synchronously each time.

**Environments:**
- Prettier Version: `1.19.1`
- Running Prettier via: `Node.js API`
- Runtime: `Node.js v12.14.0`
- Operating System: `Windows`

**Steps to reproduce:**

Call `prettier.format` API multiple times as seen in the following example.

```js
// main.js
//
// Flamegraph command: clinic flame -- node main.js

const prettier = require('prettier');

// Options are fixed
const options = {
    semi: true,
    singleQuote: true,
    trailingComma: 'es5',
    tabWidth: 4,
    printWidth: 60,
    endOfLine: 'lf',
    arrowParens: 'always',
    parser: 'typescript',
};

// Three differently named functions to show different call stacks
function foo() {
    prettier.format('', options);
}

function bar() {
    prettier.format('', options);
}

function baz() {
    prettier.format('', options);
}

// Repeatedly calling `prettier.format` with the same options
// causes plugins to be constantly reloaded from directories
// as seen in the flamegraph below.
const max = 100;
for (let i = 0; i < max; i++) {
    foo();
    bar();
    baz();
}
```

**Expected behavior:**

It should be possible to load plugins (if necessary) only once for a given set of options and then use the returned formatter to format different input texts.

**Actual behavior:**

Plugins are reloaded from directories with each call to `prettier.format` as seen in [this flamegraph](https://upload.clinicjs.org/public/66c03b913bf6053ddb6dbffc4181266671f0e98adf6a45655731583d159067b7/16704.clinic-flame.html)

**Comment by velut** [comment]
I'm sharing a temporary workaround to improve performance for API users.

Using [patch-package](https://www.npmjs.com/package/patch-package) and applying this patch

```diff
diff --git a/node_modules/prettier/index.js b/node_modules/prettier/index.js
index 96cfec8..d455497 100644
--- a/node_modules/prettier/index.js
+++ b/node_modules/prettier/index.js
@@ -44202,7 +44202,8 @@ function _withPlugins(fn) {
     var args = Array.from(arguments);
     var opts = args[1] || {};
     args[1] = Object.assign({}, opts, {
-      plugins: loadPlugins_1(opts.plugins, opts.pluginSearchDirs)
+      // WORKAROUND: Improve formatting performance
+      plugins: internalPlugins
     });
     return fn.apply(null, args);
   };
```

prevents searching directories for external plugins, improving formatting speed.
The API exported by prettier remains the same.

Using the example in my first post, total time for the formatting operations (the for loop) goes from ~1.8 seconds down to ~400ms.

**Comment by fisker** [comment]
I thought this is by design

**Comment by velut** [comment]
I tried to benchmark two possible solutions against the current `prettier.format` implementation.

1. `__formatWithoutPlugins` is similar to the solution described in my previous post. It formats text as prettier does by default, but skips searching and loading external plugins.
   Formatting options are normalized on each call.

2. `__formatWithOptions` implements the expected behavior described in my opening post. This function takes an options object, normalizes it, and returns a formatting function that can be called multiple times with a text parameter.
   External plugin search is skipped also in this solution.

Running the `benchmark.js` file below, where each formatting function is called 1000 times, I get these results on my machine (i7 6700HQ + 8GB RAM + SSD):

```
Iterations: 1000

prettier.format:
   Total Time (ms): 4306.780501000583

formatWithoutPlugins:
   Total Time (ms): 472.1005010008812

formatWithOptions:
   Total Time (ms): 38.45649900101125
```


To test my solutions, apply the following patch using [patch-package](https://www.npmjs.com/package/patch-package) and then run the `benchmark.js` file below.

```diff
diff --git a/node_modules/prettier/index.js b/node_modules/prettier/index.js
index 96cfec8..4740ebf 100644
--- a/node_modules/prettier/index.js
+++ b/node_modules/prettier/index.js
@@ -44219,9 +44219,35 @@ function withPlugins(fn) {
 }

 var formatWithCursor = withPlugins(core.formatWithCursor);
+
+// WORKAROUND: Improve formatting performance
+function __normalizeOptions(opts) {
+  return normalizeOptions$1(Object.assign({}, opts, {
+    plugins: internalPlugins
+  }));
+}
+
+function __formatWithoutPlugins(text, opts) {
+  opts = __normalizeOptions(opts);
+  return format(text, opts).formatted;
+}
+
+function __formatWithOptions(opts) {
+  opts = __normalizeOptions(opts);
+  return function(text) {
+    // Need to clone, otherwise formatted text is wrong
+    var optsClone = Object.assign({}, opts);
+    return format(text, optsClone).formatted;
+  }
+}
+
 var src$1 = {
   formatWithCursor,

+  // WORKAROUND: Improve formatting performance
+  __formatWithoutPlugins,
+  __formatWithOptions,
+
   format(text, opts) {
     return formatWithCursor(text, opts).formatted;
   },
```

`benchmark.js`

```js
// benchmark.js

const { performance } = require('perf_hooks');

const {
    format,
    __formatWithoutPlugins,
    __formatWithOptions,
} = require('prettier');

const max = 1000;

// Options are fixed.
const options = {
    semi: true,
    singleQuote: true,
    trailingComma: 'es5',
    tabWidth: 4,
    printWidth: 60,
    endOfLine: 'lf',
    arrowParens: 'always',
    parser: 'typescript',
};

// Default `prettier.format` function.
function defaultFormat() {
    const start = performance.now();
    for (let i = 0; i < max; i++) {
        format('', options);
    }
    const end = performance.now();
    const time = end - start;
    console.log('\nprettier.format:');
    console.log(`   Total Time (ms): ${time}`);
}

// Format without loading external plugins,
// but normalize options on each call.
function formatWithoutPlugins() {
    const start = performance.now();
    for (let i = 0; i < max; i++) {
        __formatWithoutPlugins('', options);
    }
    const end = performance.now();
    const time = end - start;
    console.log('\nformatWithoutPlugins:');
    console.log(`   Total Time (ms): ${time}`);
}

// Create a custom formatter and normalize options once,
// do not load external plugins.
function formatWithOptions() {
    const myFormat = __formatWithOptions(options);
    const start = performance.now();
    for (let i = 0; i < max; i++) {
        myFormat('');
    }
    const end = performance.now();
    const time = end - start;
    console.log('\nformatWithOptions:');
    console.log(`   Total Time (ms): ${time}`);
}

console.log(`Iterations: ${max}`);
defaultFormat();
formatWithoutPlugins();
formatWithOptions();
```

**Comment by thorn0** [comment]
Unless you're formatting only very short expressions, the overhead looks negligible if compared to how long the actual formatting takes .

But still I think we should cache the plugins the same way we're caching resolved config files and add `clearPluginsCache` to the API.

**Comment by thorn0** [comment]
closed by #7485
